### ISSUE TITLE
Is there a way to spawn multiple workers at once to render different PDFs?

### ISSUE THREAD

**Issue by c-x-g** [mozilla/pdf.js#16871]
Attach (recommended) or Link to PDF file here:

Configuration:
- Web browser and its version: FireFox 116.0.3-1
- Operating system and its version: Arch Linux 6.1.46-1-lts
- PDF.js version: 3.9.179
- Is a browser extension: No

Steps to reproduce the problem:
1. I'm trying to render images of PDFs on a webpage and with my current implementation they can only be rendered in single-file synchronously; please see GIF below for an illustration:

![Peek 2023-08-24 11-12](https://github.com/mozilla/pdf.js/assets/5555019/ef143559-a418-4a64-9625-160c3f47bb96)

What is the expected behavior? (add screenshot)
Desired behavior is rendering of images asynchronously so that the bottleneck is the most detailed PDF image to be rendered

What went wrong? (add screenshot)
Loading is slower than I like, particularly if eventually I may want to load a lot more PDFs, is there a way to spawn simultaneous workers to speed up the process?

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension):

**Comment by Snuffleupagus** [comment]
> [...] please see GIF below for an illustration:

Unfortunately that's not really helpful, please see https://github.com/mozilla/pdf.js/blob/master/.github/CONTRIBUTING.md (emphasis mine):
> If you are developing a custom solution, first check the examples at https://github.com/mozilla/pdf.js#learning and search existing issues. If this does not help, **please prepare a short well-documented example that demonstrates the problem and make it accessible online** on your website, JS Bin, GitHub, etc. before opening a new issue or contacting us in the Matrix room -- **keep in mind that just code snippets won't help us troubleshoot the problem**.

---

> Loading is slower than I like, particularly if eventually I may want to load a lot more PDFs, is there a way to spawn simultaneous workers to speed up the process?

By default, unless you manually set certain options, each `getDocument`-call will use its own worker-thread.

**Comment by c-x-g** [comment]
Thanks for the reply, yesterday I ran into an issue initially with asynchronous code where the browser would give an error: `Cannot use the same canvas during multiple render() operations`; possibly the same issue [here](https://github.com/mozilla/pdf.js/issues/10018)

After reading some docs it appeared that if a worker was currently in the rendering process of a PDF, it would need to be waited upon for completion before the next PDF could begin. Unfortunately I don't think I can reproduce this issue now which is strange, but I implemented a solution afterward to force an array of file names to render one after another using callbacks and this is where the performance bottleneck for this issue I raised came about.

I removed the callback logic just now and this time I no longer got the issue from the first paragraph above and it appears to be working as expected, I'm not sure what I did wrong previously. 

![Peek 2023-08-24 11-22](https://github.com/mozilla/pdf.js/assets/5555019/d86dc3e9-eca3-4f48-a7cf-d5ed9a6b1821)
